File: solutions/python/generate_parentheses.py
# Runtime 48 ms Beats 43.35%
# Memory 16.8 MB Beats 11.7%

class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        result = []
        self.generate_parenthesis_recursive(n, n, '', result)
        return result

    def generate_parenthesis_recursive(self, left: int, right: int, current: str, result: List[str]):
        if left == 0 and right == 0:
            result.append(current)
            return
        if left > 0:
            self.generate_parenthesis_recursive(left - 1, right, current + '(', result)
        if right > left:
            self.generate_parenthesis_recursive(left, right - 1, current + ')', result)

File: solutions/csharp/GenerateParentheses.cs
// Runtime 129 ms Beats 87.29%
// Memory 45.9 MB Beats 58.67%

namespace GenerateParentheses {
    public class Solution {
        public IList<string> GenerateParenthesis(int n) {
            var result = new List<string>();
            GenerateParenthesisRecursive(n, n, "", result);
            return result;
        }

        void GenerateParenthesisRecursive(int left, int right, string current, List<string> result) {
            if (left == 0 && right == 0) {
                result.Add(current);
                return;
            }

            if (left > 0) {
                this.GenerateParenthesisRecursive(left - 1, right, current + '(', result);
            }
            if (right > left) {
                this.GenerateParenthesisRecursive(left, right - 1, current + ')', result);
            }
        }
    }
}